### ISSUE TITLE
Bulk query with large result set fails silently

### ISSUE THREAD

**Issue by stemail23** [jsforce/jsforce#162]
I'm attempting to use bulk.query with large result sets, and nothing is happening!

for a request like: bulk.query('select Id from Account') with a result set of potentially 400,000 records, I'm receiving no 'record', 'error', or 'end' events. Similar queries with a where clause filter to bring the result set down below about 100,000 records seem to work fine. 

I also tried to pipe the results straight to file, and similarly got no results for large queries.

**Comment by mlim1972** [comment]
I can confirm this behavior.  I have results with about 250k and it fails silently as described above.  I eliminated the bulk and use straight query and that works fine but would still like to use bulk since query requires the maxFetch size.

**Comment by jimrae** [comment]
I also am experiencing this same issue.

**Comment by jeffoconnell** [comment]
I ran into this too.  The straight bulk.query would die on larger result sets.  Moving to use the batch/poll strategy got me going though.

conn.login uid, pwd, (err, res) ->
  if err
    console.error(err)

  jobId = null
  batchId = null

  job = conn.bulk.createJob(objectName, 'query')
  batch = job.createBatch()
  batch.execute(querytext)

  batch.on 'queue', (msg)->
    jobId = job.id
    batchId = batch.id
    console.log "Query Job: #{jobId}, Batch: #{batchId} queued"

    batch.poll(1000, 500000) #need to make sure 2nd argument is long enough or job aborts
    return

  batch.on 'response', (resp) ->
    resp.forEach (response, i) ->
      endBatch = conn.bulk.job(jobId).batch(batchId)

      endBatch.retrieve (err, results) ->
        if err
          return console.error(err)
        i = 0
        while i < results.length
          resultId = results[i].id
          console.log "Writing to disk"
          endBatch.result(resultId).stream().pipe fs.createWriteStream('./result' + i + '.csv')
          i++
        return

    job.close
    return

**Comment by sterlingwes** [comment]
Looks like this is the [intended behaviour](https://github.com/jsforce/jsforce/blob/master/lib/api/bulk.js#L801) (only the first batch result is piped).

@jeffoconnell 's solution is therefore more complete. Thanks Jeff! Wish this was in the docs, would have saved me a lot of time.

**Comment by jimrae** [comment]
Any thoughts on how to process the result sets in the system, instead of
piping to a file? I have an application where I need to work through the
returned record set for analysis purposes in javascript.

**Comment by sterlingwes** [comment]
The stream that's getting piped is a [Transform stream](https://github.com/nodejs/readable-stream/blob/master/doc/stream.markdown#class-streamtransform) I'm pretty sure. You can read the stream line by line or possibly transform further (include any existing `_transform` method within your own). I think you'd be better off reading the written files with a proper CSV parser module. Benefit there is you have the source on disk for debug reference.

**Comment by KyleAMathews** [comment]
Yeah, banged my head against this for a while as well. @jeffoconnell's solution worked perfectly.

**Comment by KyleAMathews** [comment]
What was confusing for me is the first ~49 records would be returned then... nothing.

**Comment by gwhobbs** [comment]
I am also getting ~48 records using the bulk query method.  Any ideas why this is happening?

**Comment by sterlingwes** [comment]
Have you read the whole thread?






On Sat, Dec 19, 2015 at 12:55 PM -0800, "Grant Hobbs" <notifications@github.com> wrote:












I am also getting ~48 records using the bulk query method.  Any ideas why this is happening?

**Comment by gwhobbs** [comment]
Yes.  I am unfamiliar with piping in Node so that might be the issue.  I did try the code @jeffoconnell posted which work fine when I pipe to process.stdout, but only returns the first few results when piping to a RecordStream as is done in the code @sterlingwes referenced.  I am not sure what that points to in terms of a solution.

**Comment by gwhobbs** [comment]
After struggling with this a bit more today: the loop is making it through all the records, but the transform stream stops working after ~48 records (if a custom object is pushed down the pipe before and after the loop, the first one makes it and the second one does not).  Also, the "unpipe" event handler is never called if more than ~48 records are pushed. 

Does anyone with a better grasp on Node streams (and jsforce RecordStreams) have any idea why this would be happening?

**Comment by gwhobbs** [comment]
Changing the highWaterMark on the PassThrough stream created on line 123 of record-stream.js seems to fix the problem for me.  I tried a few things in the CSV parsing function (line 240) to more gracefully handle backpressure (the for loop after the parsing finishes is flooding the pipe) but was not successful.  This was an easy change and appears to have completely fixed my problem.

https://github.com/gwhobbs/jsforce/tree/fix-bulk-query

I submitted a pull request with the tweak: https://github.com/jsforce/jsforce/pull/320.
